File: python/src/multiagent_boilerplate/agents/run_agent.py
"""The generic agent loop every role (lead, subagent, citation, judge) runs through. Not
role-specific logic - that lives entirely in specs/prompts and specs/agents. This function is
the harness-enforced tool-use loop: call model, execute any requested tool calls through the
harness, feed results back, repeat until the model stops requesting tools."""

from __future__ import annotations

from dataclasses import dataclass

from ..cost.pricing import compute_cost_usd
from ..harness import (
    AgentRoleDef,
    AuditCorrelation,
    Harness,
    HarnessToolCall,
    RunBudget,
    ToolCallBudget,
    ToolError,
    ToolOutcome,
    classify_error,
    load_agent_role,
    load_prompt,
)
from ..providers.types import (
    ChatCompletionRequest,
    ChatModel,
    ProviderMessage,
    TextBlock,
    ToolCallBlock,
    ToolDefinition,
    ToolResultBlock,
    text_of,
)
from ..tools.types import ToolRuntime
from ..tracing.tracer import Tracer
from .review import ReviewSignals, derive_review_flags
from .types import AgentResult


def _model_facing_error(outcome: ToolOutcome) -> dict:
    """The error shape the *model* sees. Auth failures are redacted to 'not permitted' - the
    model gets no detail about why, while the audit trail keeps the real message (notes
    section 12)."""
    error = outcome.error
    assert error is not None
    message = "not permitted" if error.type == "auth" else error.message
    payload: dict = {"type": error.type, "message": message, "retryable": error.retryable}
    if error.code:
        payload["code"] = error.code
    return {"status": outcome.status, "error": payload}


@dataclass
class RunAgentParams:
    role_name: str
    user_prompt: str
    model: ChatModel
    harness: Harness
    runtime: ToolRuntime
    tracer: Tracer
    task_id: str
    delegation_depth: int
    parent_span_id: str | None = None
    extra_system_context: str | None = None
    max_turns: int | None = None
    # Shared across the whole run (lead + every subagent). When exhausted, the agent stops
    # before its next model call.
    run_budget: RunBudget | None = None


async def run_agent(params: RunAgentParams) -> AgentResult:
    role_data = load_agent_role(params.role_name)
    role = AgentRoleDef.from_dict(role_data)
    system_prompt = load_prompt(role.prompt_file)
    if params.extra_system_context:
        system_prompt = f"{system_prompt}\n\n---\n\n{params.extra_system_context}"

    budget = ToolCallBudget(role.role, role.budget["maxToolCalls"])
    max_turns = params.max_turns if params.max_turns is not None else role.budget["maxToolCalls"] + 2

    messages: list[ProviderMessage] = [ProviderMessage(role="user", content=[TextBlock(text=params.user_prompt)])]
    artifact_refs: list[dict] = []
    # Errors returned to the model but not subsequently recovered from feed the deterministic
    # needs_review derivation (notes section 16a). Tracked here, evaluated at return.
    unrecovered_errors: list[ToolError] = []

    agent_span = params.tracer.start_span(
        "agent",
        f"{role.role}:{params.task_id}",
        parent_span_id=params.parent_span_id,
        agent_role=role.role,
        delegation_depth=params.delegation_depth,
    )

    tool_defs = [
        ToolDefinition(name=d["name"], description=d["description"], input_schema=d["input_schema"])
        for d in params.harness.tool_definitions(role)
    ]

    def finish(status: str, text: str, last_stop_reason: str | None = None) -> AgentResult:
        """Builds the AgentResult and attaches the deterministically-derived review flags."""
        flags = derive_review_flags(
            ReviewSignals(
                status=status,  # type: ignore[arg-type]
                unrecovered_errors=unrecovered_errors,
                final_text=text,
                last_stop_reason=last_stop_reason,  # type: ignore[arg-type]
            )
        )
        return AgentResult(
            task_id=params.task_id,
            role=role.role,
            text=text,
            artifact_refs=artifact_refs,
            status=status,  # type: ignore[arg-type]
            needs_review=len(flags) > 0,
            review_flags=flags,
        )

    for turn in range(max_turns):
        # Session-level cost ceiling, checked before spend. Distinct from the per-agent
        # tool-call cap and delegation-depth cap - this bounds total tokens across the swarm.
        if params.run_budget is not None and params.run_budget.is_exhausted():
            params.tracer.end_span(agent_span, "partial", attributes={"stoppedReason": "run_budget_exhausted"})
            return finish("partial", "(stopped: run token budget exhausted)")

        model_span = params.tracer.start_span(
            "model_call",
            f"{role.role} turn {turn}",
            parent_span_id=agent_span.span_id,
            agent_role=role.role,
            delegation_depth=params.delegation_depth,
        )

        result = await params.model.complete(
            ChatCompletionRequest(system=system_prompt, messages=messages, tools=tool_defs)
        )
        if params.run_budget is not None:
            params.run_budget.record(result.usage)
        cost = compute_cost_usd(params.model.provider, params.model.model, result.usage)
        params.tracer.end_span(
            model_span,
            "ok",
            token_usage={"inputTokens": result.usage.input_tokens, "outputTokens": result.usage.output_tokens},
            cost_usd=cost,
        )
        messages.append(result.message)

        if result.stop_reason != "tool_use":
            params.tracer.end_span(agent_span, "ok")
            return finish("ok", text_of(result.message), result.stop_reason)

        tool_calls = [b for b in result.message.content if isinstance(b, ToolCallBlock)]
        tool_results: list[ToolResultBlock] = []

        for call in tool_calls:
            tool_span = params.tracer.start_span(
                "tool_call",
                call.name,
                parent_span_id=agent_span.span_id,
                agent_role=role.role,
                delegation_depth=params.delegation_depth,
            )

            # Exhausting the per-agent tool-call budget is a harness guarantee, not a tool
            # failure: the harness forces a terminal `rejected` outcome regardless of the model.
            try:
                budget.consume()
                outcome = await params.harness.execute(
                    role,
                    HarnessToolCall(
                        idempotency_key=call.id,
                        tool_name=call.name,
                        input=call.input,
                        delegation_depth=params.delegation_depth,
                        # trace_id / session_id share the tracer instance (one per run here);
                        # request_id is this agent invocation. Rotate the tracer per turn in a
                        # multi-turn conversation.
                        correlation=AuditCorrelation(
                            trace_id=params.tracer.trace_id,
                            session_id=params.tracer.trace_id,
                            request_id=params.task_id,
                        ),
                    ),
                    params.runtime,
                )
            except Exception as budget_error:  # noqa: BLE001 - converted to a rejected outcome
                outcome = ToolOutcome(status="rejected", error=classify_error(budget_error))

            if outcome.status == "ok":
                params.tracer.end_span(tool_span, "ok")
                if call.name == "write_artifact":
                    artifact_refs.append(outcome.output)
                tool_results.append(ToolResultBlock(tool_call_id=call.id, output=outcome.output))
            else:
                assert outcome.error is not None
                unrecovered_errors.append(outcome.error)
                params.tracer.end_span(
                    tool_span,
                    "error",
                    attributes={
                        "errorType": outcome.error.type,
                        "retryable": outcome.error.retryable,
                        "code": outcome.error.code,
                    },
                )
                tool_results.append(
                    ToolResultBlock(tool_call_id=call.id, output=_model_facing_error(outcome), is_error=True)
                )

        messages.append(ProviderMessage(role="tool", content=list(tool_results)))

    params.tracer.end_span(agent_span, "partial")
    return finish("partial", "(stopped: exceeded max turns before reaching a final answer)")

File: typescript/src/agents/runAgent.ts
import type { ChatModel, ContentBlock, ProviderMessage } from "../providers/types.js";
import type { Harness } from "../harness/index.js";
import { RunBudget, ToolCallBudget, classifyError, loadAgentRole, loadPrompt, type ToolError, type ToolOutcome } from "../harness/index.js";
import type { AgentRoleDef } from "../harness/scope.js";
import type { ToolRuntime } from "../tools/types.js";
import type { Tracer } from "../tracing/tracer.js";
import { computeCostUsd } from "../cost/pricing.js";
import { deriveReviewFlags } from "./review.js";
import type { AgentResult } from "./types.js";

/**
 * The error shape the *model* sees. Auth failures are redacted to "not permitted" - the model
 * gets no detail about why, while the audit trail keeps the real message (notes section 12).
 */
function modelFacingError(outcome: Extract<ToolOutcome, { status: "error" | "rejected" }>): Record<string, unknown> {
  const { type, retryable, code } = outcome.error;
  const message = type === "auth" ? "not permitted" : outcome.error.message;
  return { status: outcome.status, error: { type, message, retryable, ...(code ? { code } : {}) } };
}

export interface RunAgentParams {
  roleName: string;
  userPrompt: string;
  model: ChatModel;
  harness: Harness;
  runtime: ToolRuntime;
  tracer: Tracer;
  taskId: string;
  delegationDepth: number;
  parentSpanId?: string | null;
  /** Appended to the role's base system prompt, e.g. "Objective: ...\nOutput format: ..." for a subagent's specific task. */
  extraSystemContext?: string;
  /** Caps how many tool-use turns the loop will take even if the budget config is larger - a hard backstop against infinite loops. */
  maxTurns?: number;
  /** Shared across the whole run (lead + every subagent). When exhausted, the agent stops before its next model call. */
  runBudget?: RunBudget;
}

/**
 * The generic agent loop every role (lead, subagent, citation, judge) runs through. Not
 * role-specific logic - that lives entirely in specs/prompts and specs/agents. This function
 * is the harness-enforced tool-use loop: call model, execute any requested tool calls through
 * the harness, feed results back, repeat until the model stops requesting tools.
 */
export async function runAgent(params: RunAgentParams): Promise<AgentResult> {
  const role = loadAgentRole(params.roleName) as unknown as AgentRoleDef & { promptFile: string };
  const systemPrompt = loadPrompt(role.promptFile) + (params.extraSystemContext ? `\n\n---\n\n${params.extraSystemContext}` : "");
  const budget = new ToolCallBudget(role.role, role.budget.maxToolCalls);
  const maxTurns = params.maxTurns ?? role.budget.maxToolCalls + 2;

  const messages: ProviderMessage[] = [{ role: "user", content: [{ type: "text", text: params.userPrompt }] }];
  const artifactRefs: AgentResult["artifactRefs"] = [];
  // Errors returned to the model but not subsequently recovered from feed the deterministic
  // needs_review derivation (notes section 16a). Tracked here, evaluated at return.
  const unrecoveredErrors: ToolError[] = [];

  const agentSpan = params.tracer.startSpan("agent", `${role.role}:${params.taskId}`, {
    parentSpanId: params.parentSpanId ?? null,
    agentRole: role.role,
    delegationDepth: params.delegationDepth,
  });

  // Builds the AgentResult and attaches the deterministically-derived review flags.
  const finish = (
    status: AgentResult["status"],
    text: string,
    lastStopReason?: "end_turn" | "tool_use" | "max_tokens",
  ): AgentResult => {
    const reviewFlags = deriveReviewFlags({ status, unrecoveredErrors, finalText: text, ...(lastStopReason ? { lastStopReason } : {}) });
    return { taskId: params.taskId, role: role.role, text, artifactRefs, status, needsReview: reviewFlags.length > 0, reviewFlags };
  };

  for (let turn = 0; turn < maxTurns; turn++) {
    // Session-level cost ceiling, checked before spend. Distinct from the per-agent tool-call
    // cap and the delegation-depth cap - this bounds total tokens across the whole swarm.
    if (params.runBudget?.isExhausted()) {
      params.tracer.endSpan(agentSpan, "partial", { attributes: { stoppedReason: "run_budget_exhausted" } });
      return finish("partial", "(stopped: run token budget exhausted)");
    }

    const modelSpan = params.tracer.startSpan("model_call", `${role.role} turn ${turn}`, {
      parentSpanId: agentSpan.spanId,
      agentRole: role.role,
      delegationDepth: params.delegationDepth,
    });

    const result = await params.model.complete({
      system: systemPrompt,
      messages,
      tools: params.harness.toolDefinitions(role),
    });
    params.runBudget?.record(result.usage);
    const costUsd = computeCostUsd(params.model.provider, params.model.model, result.usage);
    params.tracer.endSpan(modelSpan, "ok", { tokenUsage: result.usage, costUsd });
    messages.push(result.message);

    if (result.stopReason !== "tool_use") {
      params.tracer.endSpan(agentSpan, "ok");
      return finish("ok", textOf(result.message), result.stopReason);
    }

    const toolCalls = result.message.content.filter((b): b is Extract<ContentBlock, { type: "tool_call" }> => b.type === "tool_call");
    const toolResults: ContentBlock[] = [];

    for (const call of toolCalls) {
      const toolSpan = params.tracer.startSpan("tool_call", call.name, {
        parentSpanId: agentSpan.spanId,
        agentRole: role.role,
        delegationDepth: params.delegationDepth,
      });

      // Exhausting the per-agent tool-call budget is a harness guarantee, not a tool failure:
      // the harness forces a terminal `rejected` outcome regardless of what the model wants next.
      let outcome: ToolOutcome;
      try {
        budget.consume();
        outcome = await params.harness.execute(
          role,
          {
            idempotencyKey: call.id,
            toolName: call.name,
            input: call.input,
            delegationDepth: params.delegationDepth,
            // trace_id / session_id share the tracer instance (one per run here); request_id is
            // this agent invocation. In a multi-turn conversation you'd rotate the tracer per turn.
            correlation: { traceId: params.tracer.traceId, sessionId: params.tracer.traceId, requestId: params.taskId },
          },
          params.runtime,
        );
      } catch (budgetError) {
        outcome = { status: "rejected", error: classifyError(budgetError) };
      }

      if (outcome.status === "ok") {
        params.tracer.endSpan(toolSpan, "ok");
        if (call.name === "write_artifact") artifactRefs.push(outcome.output as AgentResult["artifactRefs"][number]);
        toolResults.push({ type: "tool_result", toolCallId: call.id, output: outcome.output });
      } else {
        unrecoveredErrors.push(outcome.error);
        params.tracer.endSpan(toolSpan, "error", {
          attributes: { errorType: outcome.error.type, retryable: outcome.error.retryable, code: outcome.error.code ?? null },
        });
        toolResults.push({ type: "tool_result", toolCallId: call.id, output: modelFacingError(outcome), isError: true });
      }
    }

    messages.push({ role: "tool", content: toolResults });
  }

  params.tracer.endSpan(agentSpan, "partial");
  return finish("partial", "(stopped: exceeded max turns before reaching a final answer)");
}

function textOf(message: ProviderMessage): string {
  return message.content
    .filter((b): b is Extract<ContentBlock, { type: "text" }> => b.type === "text")
    .map((b) => b.text)
    .join("\n");
}
